### PR TITLE
Fix #376 --path option of install cli not working as expected

### LINKED ISSUE #376: --path feature for install cli not working as expected
--path feature for install cli not working as expected
Find issue details : https://gist.github.com/naina-verma/e03bdfb7bc60adaa22b2dbb75e48f604
tested on latest vsm 
`vagrant-registration (1.2.3)
  Version Constraint: 1.2.3
vagrant-service-manager (1.3.0.dev)
   Version Constraint: 1.3.0.dev
vagrant-sshfs (1.2.0)`

### PR DESCRIPTION
**Pull request by brgnepal**
Fix #376
- install-cli command will fail if invalid --path is given or
  path needs permission

### CODE REVIEW

**Comment by brgnepal** [comment]
@hferentschik CI is passing :+1:

File: features/install-cli.feature [hunk 1]
```diff
@@ -62,6 +62,14 @@ Feature: Command behavior of client side tools installation
     Then the exit status should be 0
     And the binary should be installed in path "#{ENV['VAGRANT_HOME']}/docker"
 
+    When I evaluate and run `bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/unknown_dir/docker`
+    Then the exit status should be 126
+    And stderr from evaluating and running "bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/unknown_dir/docker" should match /Directory path #{ENV['VAGRANT_HOME']}/unknown_dir is invalid or doesn't exist/
```
Comment (hferentschik) [inline features/install-cli.feature:70]: Nice 👍

Comment (hferentschik) [reply features/install-cli.feature:70]: What's about testing/verifying that the error message in the original bug report is not in the output? Neither here, nor in any of the successful cases of install-cli

**Comment by brgnepal** [comment]
retest this please

**Comment by brgnepal** [comment]
@hferentschik The tests seems to be passing now. This time these are passing genuinely :smile:. However, there is one testcase which is failing in terms of openshift
`bundle exec vagrant service-manager install-cli openshift` which I don't know. I tried a lot for its debugging.

However, its cli-version `bundle exec vagrant service-manager install-cli openshift --cli-version 1.3.1` is passing.
This PR has --cli-version.

The failed one is here https://ci.centos.org/job/vagrant-service-manager-budh/11/console.

However, no such issue in case of `cdk-openshift.feature` (https://github.com/budhrg/vagrant-service-manager/blob/e81d07a3e4b73042a1f633391318696e6e7750f0/features/cdk-openshift.feature).

I believe if you have not any other review, we can merge it and investigate the edge case in this issue https://github.com/projectatomic/vagrant-service-manager/issues/415

Let me know what is your opinion.

**Comment by brgnepal** [comment]
retest this please

File: features/adb-openshift.feature [hunk 2]
```diff
@@ -30,6 +34,11 @@ Feature: Command output from various OpenShift related commands in ADB
     ==> default: OpenShift service configured successfully...
     """
 
+    When I successfully run `sleep 20`
```
Comment (hferentschik) [inline features/adb-openshift.feature]: That's a pretty long timeout. Really imo this also an issue of the underlying service setup. I expect that when Vagrant up has completed, everything is up and running, but you seem to imply this is not the case.

File: features/adb-openshift.feature [hunk 3]
```diff
@@ -30,6 +34,11 @@ Feature: Command output from various OpenShift related commands in ADB
...
+    When I successfully run `bundle exec vagrant service-manager status`
```
Comment (hferentschik) [inline features/adb-openshift.feature]: "And I sucessfully" ...

File: features/adb-openshift.feature [hunk 4]
```diff
@@ -30,6 +34,11 @@ Feature: Command output from various OpenShift related commands in ADB
...
+    Then stdout from "bundle exec vagrant service-manager status" should contain "openshift - running"
+
+    When I successfully run `sleep 20`
```
Comment (hferentschik) [inline features/adb-openshift.feature]: Do you have an idea why this is necessary?

Comment (hferentschik) [reply features/adb-openshift.feature]: Also, I think we are better of to create our own step definition. Something like "When I sleep 20 seconds". Then in the implementation you are calling Ruby's sleep method. This way you are doing things within Ruby. Using `sleep 2-` effectively calls the shells sleep command which might or might not work on Windows.

File: features/adb-openshift.feature [hunk 5]
```diff
@@ -53,23 +62,23 @@ Feature: Command output from various OpenShift related commands in ADB
...
+    When I successfully run `bundle exec vagrant service-manager install-cli openshift --cli-version 1.3.0`
     Then the exit status should be 0
     And the binary "oc" of service "openshift" should be installed with version "1.3.0"
+    And stdout from "bundle exec vagrant service-manager install-cli openshift --cli-version 1.3.0" should be evaluable in a shell
 
-    When I evaluate and run `bundle exec vagrant service-manager install-cli openshift --path #{ENV['VAGRANT_HOME']}/oc`
+    When I evaluate and run `bundle exec vagrant service-manager install-cli openshift --cli-version 1.3.0 --path #{ENV['VAGRANT_HOME']}/oc`
     Then the exit status should be 0
     And the binary should be installed in path "#{ENV['VAGRANT_HOME']}/oc"
+    And stdout from "bundle exec vagrant service-manager install-cli openshift --cli-version 1.3.0 --path #{ENV['VAGRANT_HOME']}/oc" should be evaluable in a shell
```
Comment (hferentschik) [inline features/adb-openshift.feature]: 👍

File: features/cdk-openshift.feature [hunk 6]
```diff
@@ -18,11 +18,23 @@ Feature: Command output from various OpenShift related commands in CDK
...
+      end
+
       config.servicemanager.services = 'docker, openshift'
     end
     """
 
     When I successfully run `bundle exec vagrant up --provider <provider>`
+    Then stdout from "bundle exec vagrant up --provider <provider>" should contain:
+    """
+    ==> default: Docker service configured successfully...
```
Comment (hferentschik) [inline features/cdk-openshift.feature:32]: 👍

File: features/install-cli.feature [hunk 7]
```diff
@@ -53,14 +53,25 @@ Feature: Command behavior of client side tools installation
...
+    And the stderr should not contain anything
 
     When I evaluate and run `bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/docker`
     Then the exit status should be 0
     And the binary should be installed in path "#{ENV['VAGRANT_HOME']}/docker"
+    And the stderr should not contain anything
+
+    When I evaluate and run `bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/unknown_dir/docker`
+    Then the exit status should be 126
+    And stderr from evaluating and running "bundle exec vagrant service-manager install-cli docker --path #{ENV['VAGRANT_HOME']}/unknown_dir/docker" should match /Directory path #{ENV['VAGRANT_HOME']}/unknown_dir is invalid or doesn't exist/
```
Comment (hferentschik) [inline features/install-cli.feature:70]: 👍

File: features/support/env.rb [hunk 8]
```diff
@@ -93,6 +93,12 @@ def extract_process_id(data)
   aruba.command_monitor.find(Aruba.platform.detect_ruby(cmd)).send(:stdout) =~ /#{regexp}/
 end
 
+Then(%r{^stderr from evaluating and running "([^"]*)" should match /(.*)/$}) do |cmd, regexp|
```
Comment (hferentschik) [inline features/support/env.rb:101]: :-)

File: lib/vagrant-service-manager/installer.rb [hunk 9]
```diff
@@ -23,21 +24,30 @@ def install
           @binary_handler.install
         end
 
-        @binary_handler.print_message
+        @binary_handler.print_message(@option_msg)
+      end
+
+      def format_options
+        msg = ''
```
Comment (hferentschik) [inline lib/vagrant-service-manager/installer.rb:31]: Is this to handle the right output for the 'eval'?

**Review by hferentschik — CHANGES_REQUESTED** [review]
Getting there. Looking quite good now. Not quite sure what to think about 20 seconds timeout. Does it really have to be so high. We also needs to get to the bottom of this from a CDK side of things. There should be no wait required really. 

For me to get through the test suite, I still need https://github.com/hferentschik/vagrant-service-manager/commit/19cf82481a843d9f4e16ad6421bcb3dde1132028

**Comment by hferentschik** [comment]
retest

**Comment by hferentschik** [comment]
retest

**Comment by hferentschik** [comment]
test this please

File: lib/vagrant-service-manager/installer.rb (hunk 9 above)
Comment (brgnepal) [reply lib/vagrant-service-manager/installer.rb:31]: @hferentschik Yes

**Comment by brgnepal** [comment]
@hferentschik Hi, I have updated the tests now. It include both evalauable the install-cli command and added step for sleep too.

**Comment by brgnepal** [comment]
retest this please

**Comment by hferentschik** [comment]
test this please

File: features/adb-openshift.feature [hunk 10]
```diff
@@ -19,6 +19,10 @@ Feature: Command output from various OpenShift related commands in ADB
       config.vm.network :private_network, ip: '<ip>'
       config.vm.synced_folder '.', '/vagrant', disabled: true
 
+      config.vm.provider "libvirt" do |v, _override|
+        v.memory = 3072
+      end
+
```
Comment (hferentschik) [inline features/adb-openshift.feature:24]: Why don't you add the virtualbox memory settings here as well?

**Comment by hferentschik** [comment]
@budhrg Did you create an issue somewhere in adb to find out why we need a wait period before between 'vagrant up' and the service-manager 'status'? That does not seem right to me. 

Also, lets create a follow up issue in service-manager to find a more direct condition than sleeping 10 seconds. This might or might not work depending on the machine the tests are running. It could be something like "And wait max x seconds to OpenShift to be operational" where the implementation checks in a loop some condition of the OpenShift cluster and times out after max seconds. Adding random waits to a test is a bad smell.

**Comment by hferentschik** [comment]
@budhrg - does this `bundle exec rake clean features BOX=adb,cdk` pass all tests on your work station?

**Comment by hferentschik** [comment]
@budhrg Didn't we want to enable the Kubernates features on CI as well?

**Review by hferentschik — CHANGES_REQUESTED** [review]
The tests are not passing! You need to convert the captured string into an int to make the sleep work.

```
      can't convert String into time interval (TypeError)
  /root/vagrant-service-manager/feature/support/env.rb:188:in `sleep'
```

**Comment by hferentschik** [comment]
@budhrg For what it's worth `bundle exec rake clean features BOX=adb,cdk` seems to work for me now ignoring the string conversion error (note this in itself makes of course all adb openshift tests to be skipped)

File: features/adb-openshift.feature [hunk 11]
```diff
@@ -30,7 +38,12 @@ Feature: Command output from various OpenShift related commands in ADB
     ==> default: OpenShift service configured successfully...
     """
 
-    When I successfully run `bundle exec vagrant service-manager env openshift`
+    When I sleep for 10 seconds
+    And I successfully run `bundle exec vagrant service-manager status`
+    Then stdout from "bundle exec vagrant service-manager status" should contain "openshift - running"
+
+    When I sleep for 10 seconds
```
Comment (LalatenduMohanty) [inline features/adb-openshift.feature:40]: This should not be the case. We need to investigate the root cause of why we need 10s delay.

**Review by hferentschik — APPROVED** [review]
:+1

**Pull request merged**